resource/profile.rs

```diff
@@ -0,0 +1,742 @@
+            Self {
+                data: Some(CompileProfileData {
+                    stage_start: Instant::now(), // irrelevant
+                    translation: Duration::ZERO,
+                    validation: Duration::ZERO,
+                    annotation: Duration::ZERO,
+                    compilation: Duration::ZERO,
+                }),
+            }
+        } else {
+            Self { data: None }
+        }
+    }
+
+    pub fn start(&mut self) {
+        if let Some(data) = &mut self.data {
+            data.stage_start = Instant::now()
+        }
+    }
+
+    pub fn translation_finished(&mut self) {
+        if let Some(data) = &mut self.data {
+            data.translation = data.stage_start.elapsed();
+            data.stage_start = Instant::now();
+        }
+    }
+
+    pub fn validation_finished(&mut self) {
+        if let Some(data) = &mut self.data {
+            data.validation = data.stage_start.elapsed();
+            data.stage_start = Instant::now();
+        }
+    }
+
+    pub fn annotation_finished(&mut self) {
+        if let Some(data) = &mut self.data {
+            data.annotation = data.stage_start.elapsed();
+            data.stage_start = Instant::now();
+        }
+    }
+
+    pub fn compilation_finished(&mut self) {
+        if let Some(data) = &mut self.data {
+            data.compilation = data.stage_start.elapsed();
+            data.stage_start = Instant::now();
+        }
+    }
+
+    fn total_micros(&self) -> f64 {
+        match &self.data {
+            None => 0.0,
+            Some(data) => {
+                (data.translation + data.validation + data.annotation + data.compilation).as_nanos() as f64 / 1000.0
+            }
+        }
+    }
+}
+
+impl Display for CompileProfile {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match &self.data {
+            None => writeln!(f, "  Compile[enabled=false]"),
+            Some(data) => {
+                writeln!(f, "  Compile[enabled=true, total micros={}]", self.total_micros())?;
+                writeln!(f, "    translation micros: {}", data.translation.as_nanos() as f64 / 1000.0)?;
+                writeln!(f, "    validation micros: {}", data.validation.as_nanos() as f64 / 1000.0)?;
+                writeln!(f, "    annotation micros: {}", data.annotation.as_nanos() as f64 / 1000.0)?;
+                writeln!(f, "    compilation micros: {}", data.compilation.as_nanos() as f64 / 1000.0)
+            }
+        }
+    }
+}
+
+/// Record the time different stages of a query compilation take.
+/// This struct is simplified to expect that we execute exactly these steps in a fixed order with no other (significant) operations.
+#[derive(Debug)]
+struct CompileProfileData {
+    stage_start: Instant,
+    translation: Duration,
+    validation: Duration,
+    annotation: Duration,
+    compilation: Duration,
+}
+
+#[derive(Debug)]
+pub struct EncodingProfile {
+    storage_counters: StorageCounters,
+}
```

Comment:
Profiling how many storage operations it took to encode responses.
Name options were between `ResponseProfile` and `EncodingProfile`, and the latter felt more precise.

Reply:
Agree, looks good to me